### PR TITLE
Separate the console from testing

### LINKED ISSUE #18352: Action prompt in dev mode only seems to appear if tests are present for the app
## Describe the bug
With Quarkus 2.0, it's possible to interact with the app at runtime during dev mode, relaunch tests, force a restart, etc. However, this appears that this only happens if there are tests for the app. If you launch an app in dev mode that doesn't have any associated tests, no prompt is available to interact with the app.

### Expected behavior
Prompt for action should always be available even in the absence of tests.

### Actual behavior
If the app has no tests, then there is no way to interact with the app in dev mode

## To Reproduce
Any app without tests should reproduce the issue (I think).

### LINKED ISSUE #18400: Command Mode App under Development mode is NOT able to be restarted or stopped based on user input value
## Describe the bug
https://quarkus.io/blog/introducing-command-mode/#development-mode doesn't work as expected anymore in quarkus version 2.0.0.Final. 

### Expected behavior
In Command Mode + Development Mode, Server is able to be restarted or stopped based on user input value (except with Ctrl + C)

### Actual behavior
In Command Mode + Development Mode, Server is NOT able to be restarted or stopped, must stop it with Ctrl + C.

## To Reproduce

Follow example in https://quarkus.io/blog/introducing-command-mode/#development-mode 

### Configuration
```properties
# Add your application.properties here, if applicable.

```

### Screenshots
(If applicable, add screenshots to help explain your problem.)

### Quarkus version or git rev
quarkus version 2.0.0.Final

## Additional context
It works in 1.9.2.Final.

### LINKED ISSUE #18594: Dev mode console handler does not get installed if there are no tests (picocli app)
## Describe the bug
When running my (Pico)CLI app (https://github.com/gunnarmorling/kc-cli) in Dev mode, it doesn't show me the options for restarting that CLI app, editing the passed parameters, etc. It just hangs there:

```
mvn compile quarkus:dev  -Dquarkus.args='info'

...

Quarkus application exited with code 0
```

### Expected behavior
There should be the line telling me about the dev mode options like restarting, etc.

### Actual behavior
That line isn't displayed.

## To Reproduce

Clone the project and run it in dev mode.

### Configuration
n.a.

### Screenshots
n.a.

### Output of `uname -a` or `ver`
Darwin Kernel Version 19.6.0

### Output of `java -version`
build 17-ea+23-2064

### GraalVM version (if different from Java)
21.1.0

### Quarkus version or git rev
2.0.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)
Maven 3.8.1

## Additional context
n.a.

### LINKED ISSUE #18719: Running a CLI app in quarkus dev mode (maven), does not re-run the app after a code change. (macOS
### Describe the bug

I run quarkus dev mode for a sample CLI application, and after updating the source code I press [Enter] on the terminal as suggested by the command line reference guide, and the cli app is not re-run.

OS: macOS Big Sur, version 11.3.1
Terminal: Default macOS Terminal, Version 2.11 (440)
Shell: zsh 5.8 (x86_64-apple-darwin20.0)
Quarkus: 2.0.1.Final

In addition to using `quarkus dev`, I tried directly running `./mvnw quarkus:dev`, running `bash` instead of `zsh`, generating the example with `gradle` instead of `maven`, and manually updating the generated `pom.xml` (on the maven version) to quarkus `2.0.2.Final`. All scenarios I tested worked the same way on my system.

### Expected behavior

Quarkus Dev Mode should have re-run my sample CLI app after pressing [Enter]

### Actual behavior

Only a blank line is printed on the terminal after pressing [Enter]

### How to Reproduce?

Steps to reproduce:

1. Make sure Quarkus CLI is fully updated: `jbang app install --fresh --force quarkus@quarkusio`
2. On a terminal, create a CLI app and run it in dev mode: `quarkus create cli -a mycli; cd mycli; quarkus dev`
3. On a separate terminal, update the sample code: `sed -i '' 's/, go go commando//g' mycli/src/main/java/org/acme/GreetingCommand.java`
4. Back on the first terminal, press [Enter]. This only shows an empty line. Quarkus Dev Tool never re-runs the app.

### Output of `uname -a` or `ver`

Darwin rrivas-mac 20.4.0 Darwin Kernel Version 20.4.0: Thu Apr 22 21:46:47 PDT 2021; root:xnu-7195.101.2~1/RELEASE_X86_64 x86_64

### Output of `java -version`

openjdk version "11.0.11" 2021-04-20 OpenJDK Runtime Environment AdoptOpenJDK-11.0.11+9 (build 11.0.11+9) OpenJDK 64-Bit Server VM AdoptOpenJDK-11.0.11+9 (build 11.0.11+9, mixed mode)

### GraalVM version (if different from Java)

N/A

### Quarkus version or git rev

2.0.1.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)

### Additional information

In one of the multiple times I tested, I got the following exception after stopping the dev mode with Ctl+C (many other times before and after I did not get such error, so it may or may not be related):

Exception in thread "main" java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.maven.wrapper.BootstrapMainStarter.start(BootstrapMainStarter.java:39)
	at org.apache.maven.wrapper.WrapperExecutor.execute(WrapperExecutor.java:122)
	at org.apache.maven.wrapper.MavenWrapperMain.main(MavenWrapperMain.java:61)
Caused by: java.io.IOError: java.io.IOException: Error executing 'stty icrnl echo icanon pendin dsusp ^Y lnext ^V': stty: tcsetattr: Input/output error

### PR DESCRIPTION
**Pull request by stuartwdouglas**
The console that allows you to enter commands was very tied into the
testing code, this seperates it out.

This means projects without tests can take advantage of these features.

Fixes #18594
Fixes #18352
Fixes #18400
Fixes #18719

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
Screenshot from 2021-07-15 12-19-20

**Comment by stuartwdouglas** [comment]
This also allows tests to be run against a CLI app that has exited, so you can use continuous testing as normal.

File: core/deployment/src/main/java/io/quarkus/deployment/console/AeshConsole.java [hunk 1]
```diff
@@ -163,27 +182,27 @@ private void deadlockSafeWrite() {
...
-                                    System.exit(0);
-                                }
-                            }).start();
-                            break;
-                    }
-                });
                 // Keyboard handling
                 conn.setStdinHandler(keys -> {
-                    InputHolder handler = inputHandlers.peek();
+                    var handler = inputHandler;
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/console/AeshConsole.java:203]: First instance of `var` I've seen in the Quarkus codebase :)

Comment (stuartwdouglas) [reply core/deployment/src/main/java/io/quarkus/deployment/console/AeshConsole.java:203]: I only realized I could do it a few weeks ago, and I have put a few in :-)

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/LaunchModeBuildItem.java [hunk 2]
```diff
@@ -57,4 +60,12 @@ public boolean isAuxiliaryApplication() {
     public Optional<DevModeType> getAuxiliaryDevModeType() {
         return auxiliaryDevModeType;
     }
+
+    /**
+     * If this is a test. Dev mode tests don't launch with a launch mode TEST, so this
+     * can be used to determine if we are in a dev mode test.
+     */
+    public boolean isTest() {
```
Comment (geoand) [inline core/deployment/src/main/java/io/quarkus/deployment/builditem/LaunchModeBuildItem.java:68]: This is kinda confusing to be honest. Maybe we can rename this to something more descriptive?

File: independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java [hunk 3]
```diff
@@ -369,6 +369,10 @@ public boolean isFlatClassPath() {
         return flatClassPath;
     }
 
+    public boolean isTest() {
```
Comment (geoand) [inline independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/app/QuarkusBootstrap.java:372]: Same as my previous comment on the name

**Review by geoand — COMMENTED** [review]
I can't really tell if this will create problems or not, but I do think the split and associated code cleanup is really good

**Review by geoand — APPROVED** [review]
I added some small comment that are not blocking, so approving to make sure we get this in `2.1`

**Comment by maxandersen** [comment]
should we not put this into 2.1 ?

**Comment by metacosm** [comment]
I'd like to get this into 2.1 if possible.

**Comment by geoand** [comment]
As shown by the milestone, it will be in the final version

**Comment by maxandersen** [comment]
i could have sworn that said 2.2 when I saw it at 1 am.

anyhow still `git branch --contains cdadeb1` says just in main and grepping for any of the changes in 2.1 branch finds no matches.

@geoand are you sure its in ?

**Comment by geoand** [comment]
I didn't check, but I have reason to doubt it is. The history of the labels and milestone are consistent with how Guillame backports stuff

**Comment by maxandersen** [comment]
yes but it ain't there:

```
git log --grep="Seperate the console from testing"
commit e3ca142a3dd65ec1ce159f9fb622a415a4d9dec3
Author: Stuart Douglas <stuart.w.douglas@gmail.com>
Date:   Tue Jul 13 14:01:11 2021 +1000

    Seperate the console from testing

    The console that allows you to enter commands was very tied into the
    testing code, this seperates it out.

    This means projects without tests can take advantage of these features.

    Fixes #18594
    Fixes #18352
❯ git co 2.
error: pathspec '2.' did not match any file(s) known to git
❯ git co 2.1

Switched to branch '2.1'
Your branch is up to date with 'origin/2.1'.
❯
❯ git log --grep="Seperate the console from testing"
```

or its hiding really well :)

**Comment by maxandersen** [comment]
adding back triage/backport - if @gsmet can just confirm its in and its just me who somehow can't see the code just remove the label but afaics this have NOT been backported. it is only in main branch not in 2.1 branch.

**Comment by gsmet** [comment]
It had already been backported, that's why the backport label was removed.

**Comment by maxandersen** [comment]
Okey. But why was neither commit  nor code in the branch ? You mark it as backported when you done it locally first before pushing it sometime later ?

**Comment by phillip-kruger** [comment]
Hi @stuartwdouglas - this PR removed the "Enter" for new line added here: https://github.com/quarkusio/quarkus/pull/18580

Can I add that back ?

**Comment by phillip-kruger** [comment]
For in case I can : https://github.com/quarkusio/quarkus/pull/18913

**Comment by stuartwdouglas** [comment]
Oops, sorry. That was not intentional.

**Pull request merged**